File: es25/es25.go
package main

import (
	"fmt"
	"os"
)

type Point struct {
	v [4]int
	p int
	r int
}

func LoadPoint(input *os.File, i int) (Point, error) {
	p := Point{p: i, r: 0}
	_, err := fmt.Fscanf(input, "%d,%d,%d,%d\n", &p.v[0], &p.v[1], &p.v[2], &p.v[3])
	return p, err
}

func Distance(a, b *Point) int {
	d := 0
	for i := 0; i < 4; i++ {
		if a.v[i] > b.v[i] {
			d += a.v[i] - b.v[i]
		} else {
			d += b.v[i] - a.v[i]
		}
	}
	return d
}

type Sky struct {
	points []Point
}

func (s *Sky) find(i int) int {
	for s.points[i].p != i {
		next := s.points[i].p
		s.points[i].p = s.points[next].p
		i = next
	}
	return i
}

func (s *Sky) merge(i, j int) {
	i = s.find(i)
	j = s.find(j)

	if i == j {
	} else if s.points[i].r < s.points[j].r {
		s.points[i].p = j
	} else if s.points[i].r > s.points[j].r {
		s.points[j].p = i
	} else {
		s.points[j].p = i
		s.points[i].r++
	}
}

func (s *Sky) CountConstellations() int {
	for i := 0; i < len(s.points); i++ {
		for j := 0; j < i; j++ {
			if Distance(&s.points[i], &s.points[j]) <= 3 {
				s.merge(i, j)
			}
		}
	}

	count := 0
	for i, p := range s.points {
		if p.p == i {
			count++
		}
	}
	return count
}

func LoadSky(input *os.File) Sky {
	s := Sky{}

	for i := 0; ; i++ {
		p, err := LoadPoint(input, i)
		if err != nil {
			break
		}
		s.points = append(s.points, p)
	}

	return s
}

func main() {
	rv := mainInt()
	if rv != 0 {
		os.Exit(rv)
	}
}

func mainInt() int {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: %s <filename>\n", os.Args[0])
		return -1
	}

	input, err := os.Open(os.Args[1])
	if err != nil {
		fmt.Fprintf(os.Stderr, "Cannot open %s for reading\n", os.Args[1])
		return -1
	}
	defer input.Close()

	s := LoadSky(input)
	fmt.Println("Constellation count:", s.CountConstellations())

	return 0
}

File: es25/es25.py
#!/usr/bin/env python3

import sys

class Point(object):
    def __init__(self, values, index):
        self.v = values
        self.parent = index
        self.rank = 0

def distance(p1, p2):
    d = 0
    for i in range(4):
        if p1.v[i] > p2.v[i]:
            d += p1.v[i] - p2.v[i]
        else:
            d += p2.v[i] - p1.v[i]

    return d

class Sky(object):
    def __init__(self):
        self.points = []

    def load(self, f):
        i = 0
        for line in f:
            val = [int(x) for x in line.strip().split(",")]
            self.points.append(Point(val, i))
            i += 1

    def find(self, i):
        while self.points[i].parent != i:
            next = self.points[i].parent
            self.points[i].parent = self.points[next].parent
            i = next

        return i

    def merge(self, i, j):
        i = self.find(i)
        j = self.find(j)

        if i == j:
            pass
        elif self.points[i].rank < self.points[j].rank:
            self.points[i].parent = j
        elif self.points[i].rank > self.points[j].rank:
            self.points[j].parent = i
        else:
            self.points[j].parent = i
            self.points[i].rank += 1

    def count_constellations(self):
        for i in range(len(self.points)):
            for j in range(i):
                if distance(self.points[i], self.points[j]) <= 3:
                    self.merge(i, j)

        count = 0
        for i in range(len(self.points)):
            if self.points[i].parent == i:
                count += 1

        return count

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        exit(-1)

    s = Sky()
    try:
        with open(sys.argv[1], "rt") as f:
            s.load(f)
    except:
        raise
        print("Cannot open {} for reading".format(sys.argv[1]), file = sys.stderr)
        exit(-1)

    print("Constellation count: {}".format(s.count_constellations()))
